text/0000-api-evolution.md

```diff
@@ -0,0 +1,748 @@
+to extend the enum without breakage.
+
+#### Major change: adding new fields to a variant.
+
+If the enum is public, so is the full contents of all of its variants. As per
+the rules for structs, this means it is not allowed to add any new fields (which
+will automatically be public).
+
+If you wish to allow for this kind of extensibility, consider introducing a new,
+explicit struct for the variant up front.
+
+### Traits
+
+#### Major change: adding a non-defaulted item.
+
+Adding any item without a default will immediately break all trait implementations.
+
+It's possible that in the future we will allow some kind of
+"[sealing](#sealed-traits)" to say that a trait can only be used as a bound, not
+to provide new implementations; such a trait *would* allow arbitrary items to be
+added.
+
+#### Major change: any non-trivial change to item signatures.
+
+Because traits have both implementors and consumers, any change to the signature
+of e.g. a method will affect at least one of the two parties. So, for example,
+abstracting a concrete method to use generics instead might work fine for
+clients of the trait, but would break existing implementors. (Note, as above,
+the potential for "sealed" traits to alter this dynamic.)
+
+#### Minor change: adding a defaulted item.
+
+Adding a defaulted item is technically a breaking change:
+
+```rust
+trait Trait1 {}
+trait Trait2 {
+    fn foo(&self);
+}
+
+fn use_both<T: Trait1 + Trait2>(t: &T) {
+    t.foo()
+}
+```
+
+If a `foo` method is added to `Trait1`, even with a default, it would cause a
+dispatch ambiguity in `use_both`, since the call to `foo` could be referring to
+either trait.
+
+(Note, however, that existing *implementations* of the trait are fine.)
+
+According to the basic principles of this RFC, such a change is minor: it is
+always possible to annotate the call `t.foo()` to be more explicit *in advance*
+using UFCS: `<Trait2>::foo(t)`. This kind of annotation could be done
+automatically for code in dependencies (see
+[Elaborated source](#elaborated-source)). And it would also be possible to
+mitigate this problem by allowing
+[method renaming on trait import](#trait-item-renaming).
+
+While the scenario of adding a defaulted method to a trait may seem somewhat
+obscure, the exact same hazards arise with *implementing existing traits* (see
+below), which is clearly vital to allow; we apply a similar policy to both.
+
+All that said, it is incumbent on library authors to ensure that such "minor"
+changes are in fact minor in practice: if a conflict like `t.foo()` is likely to
+arise at all often in downstream code, it would be advisable to explore a
+different choice of names. More guidelines for the standard library are given
+later on.
+
+#### Minor change: adding a defaulted type parameter.
+
+As with "[Signatures in type definitions](#signatures-in-type-definitions)",
+traits are permitted to add new type parameters as long as defaults are provided
+(which is backwards compatible).
```

Comment:
I suspect that is the same problem discussed below: https://github.com/rust-lang/rfcs/pull/1105/files#r29635875